### PR TITLE
Logging fixes

### PR DESCRIPTION
**Pull request by veerajsk**
https://www.wrike.com/open.htm?id=1097473608

- Make AlicaDefaultLogger thread safe since it is the responsibility of the logging implementation

- Set the logger immediately on construction of the context so that it can be used to log messages during construction & before init() is called. This solves some obsure bugs where the logger was used before it was initialized in the context's set api's eg. setTraceFactory or the trace factory'sconstructor etc.

- Provides the ability to set the logger multiple times before the context is initialized which is consistent with other set api's

- Destroy the logger when the context is destroyed

- Add tests for custom logging & fix the test for default logging

### CODE REVIEW

File: alica_engine/src/engine/logging/AlicaDefaultLogger.cpp
Comment (veerajsk) [inline alica_engine/src/engine/logging/AlicaDefaultLogger.cpp:31]: This solves the thread safety issue because std::cout's << operator is called just once & this operator is already thread safe. The problem earlier was that there were multiple invocations of << so the logs could get interpersed if multiple threads log at the same time which is a real possibility in alica applications

**Review by bjoernschroeder — APPROVED** [review]

File: alica_engine/src/engine/AlicaContext.cpp
Comment (david-simoes-93) [inline alica_engine/src/engine/AlicaContext.cpp:31]: I would create the following structure:

    AlicaContext::AlicaContext(const AlicaContextParams& alicaContextParams)
            : AlicaContext()
            , _validTag(ALICA_CTX_GOOD)
            , _alicaContextParams(alicaContextParams)
            , _localAgentName(alicaContextParams.agentName)
            , _configRootNode(initConfig(alicaContextParams.configPaths, alicaContextParams.agentName))
            , _alicaContextParams(alicaContextParams)
            , _validTag(ALICA_CTX_GOOD)
            , _clock(std::make_unique<AlicaClock>())
            , _localAgentName(alicaContextParams.agentName)
    {
        // sets logger based on read configs
        setLogger<AlicaDefaultLogger>();
    }

    AlicaContext::AlicaContext() // private ctor
    {
         // sets INFO logger, to be used by other ctors
        AlicaLogger::set<AlicaDefaultLogger>(Verbosity::INFO, alicaContextParams.agentName)
    }

feels cleaner

Comment (david-simoes-93) [reply alica_engine/src/engine/AlicaContext.cpp:31]: if only `initConfig` needs to log stuff, perhaps don't use `Logging` there, just standard cout/cerr? Or the call itself can check whether Logger exists, and set a default one otherwise (since it seems to be related to Logging configs, makes sense that it overrides them if they don't exist yet)

File: alica_tests/src/test/test_alica_logger.cpp
Comment (rogeriofonteles) [inline alica_tests/src/test/test_alica_logger.cpp:5]: Maybe change this when PRing the change for everything? idk, seems weird to have just one with this.

**Review by rogeriofonteles — APPROVED** [review]

File: alica_engine/src/engine/AlicaContext.cpp
Comment (veerajsk) [reply alica_engine/src/engine/AlicaContext.cpp:31]: Right now it is only `initConfig()` which can use the `Logger`, but I think it's better to have the `Logger` initialized first because anyone could add code that does logging in either one of the constructors of some member variable or in the initializer list itself & then it would be a bug. `initConfig` already has this bug & its painstaking to look through the code to see if nothing else is invoking the logger. Therefore, I think the first option you suggested is better although I'm not familiar with its syntax, let me explore more

File: alica_tests/src/test/test_alica_logger.cpp
Comment (veerajsk) [reply alica_tests/src/test/test_alica_logger.cpp:5]: didn't get you

Comment (rogeriofonteles) [reply alica_tests/src/test/test_alica_logger.cpp:5]: the namespace change would be nice to have in all test files for `alica_tests`

File: alica_engine/src/engine/AlicaContext.cpp
Comment (veerajsk) [reply alica_engine/src/engine/AlicaContext.cpp:31]: looks like the first approach is not possible because the delegating constructor should be the only one in the initializer list: see https://en.cppreference.com/w/cpp/language/constructor

I also thought about outputting to cout/cerr but I guess that could be problematic if we are relying on custom logging like ROS in which case the logs to cout/cerr may not appear, therefore I decided its better to use the Logger always

File: alica_tests/src/test/test_alica_logger.cpp
Comment (veerajsk) [reply alica_tests/src/test/test_alica_logger.cpp:5]: yes, will do that eventually. Right now, we are going through one test file at a time & rewriting/improving them

File: alica_engine/src/engine/AlicaContext.cpp
Comment (david-simoes-93) [reply alica_engine/src/engine/AlicaContext.cpp:31]: What if we delay initializing `_configRootNode` into ctor's body? So we would set default logger, `initConfig`, set custom logger?

Comment (veerajsk) [reply alica_engine/src/engine/AlicaContext.cpp:31]: still leaves the problem where other constructors/initializers can end up using the logger in future.

Comment (veerajsk) [reply alica_engine/src/engine/AlicaContext.cpp:31]: How about this solution:
```
AlicaContext::AlicaContext(const AlicaContextParams& alicaContextParams)
                : AlicaContext((AlicaLogger::set<AlicaDefaultLogger>(Verbosity::INFO, alicaContextParams.agentName), true), alicaContextParams) {}

AlicaContext::AlicaContext(bool, const AlicaContextParams& alicaContextParams) : ...
```

File: alica_engine/src/engine/logging/AlicaDefaultLogger.cpp
Comment (Dale-Koenig) [inline alica_engine/src/engine/logging/AlicaDefaultLogger.cpp:38]: Is .str() needed here?

**Review by Dale-Koenig — APPROVED** [review]

File: alica_engine/src/engine/logging/AlicaDefaultLogger.cpp
Comment (veerajsk) [reply alica_engine/src/engine/logging/AlicaDefaultLogger.cpp:38]: does not compile without this

File: alica_engine/src/engine/AlicaContext.cpp
Comment (veerajsk) [reply alica_engine/src/engine/AlicaContext.cpp:31]: @Dale-Koenig since you approved, I'm assuming you are ok with the current state here?

Comment (veerajsk) [reply alica_engine/src/engine/AlicaContext.cpp:31]: because @david-simoes-93 & me could not reach a conclusion, so we decided to leave it up to other reviewers

Comment (Dale-Koenig) [reply alica_engine/src/engine/AlicaContext.cpp:31]: Yeah no strong opinion

**Pull request merged**
